mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxExtensions.kt

```diff
@@ -48,91 +47,37 @@ inline fun <T, reified VM : BaseMvRxViewModel<S>, reified S : MvRxState> T.activ
     viewModelClass: KClass<VM> = VM::class,
     noinline keyFactory: () -> String = { viewModelClass.java.name }
 ) where T : Fragment, T : MvRxView = lifecycleAwareLazy(this) {
-    val stateFactory: () -> S = { _activityViewModelInitialStateProvider(keyFactory) }
     if (requireActivity() !is MvRxViewModelStoreOwner) throw IllegalArgumentException("Your Activity must be a MvRxViewModelStoreOwner!")
-    MvRxViewModelProvider.get(viewModelClass.java, requireActivity(), keyFactory(), stateFactory)
+    MvRxViewModelProvider.get(viewModelClass.java, S::class.java, ActivityViewModelContext(requireActivity(), _activityArgsProvider(keyFactory)), keyFactory())
         .apply { subscribe(this@activityViewModel, subscriber = { postInvalidate() }) }
 }
 
 /**
  * For internal use only. Public for inline.
  *
- * Looks for [MvRx.KEY_ARG] on the arguments of the fragment receiver to create an instance of the State class.
- * The state class must have a matching single arg constructor.
+ * Looks for [MvRx.KEY_ARG] on the arguments of the fragments.
+ */
+@Suppress("FunctionName")
+fun <T : Fragment> T._fragmentArgsProvider(): Any? = arguments?.get(MvRx.KEY_ARG)
+
+/**
+ * For internal use only. Public for inline.
  *
- * If no MvRx fragment args exist it attempts to use a empty constructor. Otherwise an exception is thrown.
+ * Looks for [MvRx.KEY_ARG] on the arguments of the fragment receiver.
  *
  * Also adds the fragment's MvRx args to the host Activity's [MvRxViewModelStore] so that they can be used to recreate initial state
  * in a new process.
  */
 @Suppress("FunctionName")
-inline fun <reified S : MvRxState, T : Fragment> T._activityViewModelInitialStateProvider(keyFactory: () -> String): S {
+inline fun <T : Fragment> T._activityArgsProvider(keyFactory: () -> String): Any? {
```

Comment:
the idea of "activity args" is confusing to me since the args are still coming from a fragment, and the function receiver is a fragment

Reply:
Agreed, let me explain the reasons a bit more in the comments.